support-models/src/main/scala/com/gu/support/workers/states/SendThankYouEmailState.scala

```diff
@@ -1,37 +1,100 @@
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: DigitalPack,
+    paymentMethod: PaymentMethod,
+    paymentSchedule: PaymentSchedule,
+    promoCode: Option[PromoCode],
+    accountNumber: String,
+    subscriptionNumber: String,
+  ) extends SendThankYouEmailDigitalSubscriptionState
+
+  case class SendThankYouEmailDigitalSubscriptionGiftPurchaseState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: DigitalPack,
+    giftRecipient: DigitalSubscriptionGiftRecipient,
+    giftCode: GeneratedGiftCode,
+    lastRedemptionDate: LocalDate,
+    paymentMethod: PaymentMethod,
+    paymentSchedule: PaymentSchedule,
+    promoCode: Option[PromoCode],
+    accountNumber: String,
+  ) extends SendThankYouEmailDigitalSubscriptionState
+
+  case class SendThankYouEmailDigitalSubscriptionCorporateRedemptionState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: DigitalPack,
+    subscriptionNumber: String,
+  ) extends SendThankYouEmailDigitalSubscriptionState
+
+  case class SendThankYouEmailDigitalSubscriptionGiftRedemptionState( //tbc
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: DigitalPack,
+  ) extends SendThankYouEmailDigitalSubscriptionState
+
+  case class SendThankYouEmailPaperState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: Paper,
+    paymentMethod: PaymentMethod,
+    paymentSchedule: PaymentSchedule,
+    promoCode: Option[PromoCode],
+    accountNumber: String,
+    subscriptionNumber: String,
+    firstDeliveryDate: LocalDate,
+  ) extends SendThankYouEmailState
+
+  case class SendThankYouEmailGuardianWeeklyState(
+    user: User,
+    salesForceContact: SalesforceContactRecord,
+    product: GuardianWeekly,
+    giftRecipient: Option[WeeklyGiftRecipient],
+    paymentMethod: PaymentMethod,
+    paymentSchedule: PaymentSchedule,
+    promoCode: Option[PromoCode],
+    accountNumber: String,
+    subscriptionNumber: String,
+    firstDeliveryDate: LocalDate,
+  ) extends SendThankYouEmailState
+
+  implicit val codec = DiscriminatedType.codec[SendThankYouEmailState](List(
+    Variant.codec[SendThankYouEmailContributionState]("Contribution", "productType"),
+    Variant.codec[SendThankYouEmailDigitalSubscriptionDirectPurchaseState]("DigitalSubscriptionDirectPurchase", "productType"),
+    Variant.codec[SendThankYouEmailDigitalSubscriptionGiftPurchaseState]("DigitalSubscriptionGiftPurchase", "productType"),
+    Variant.codec[SendThankYouEmailDigitalSubscriptionCorporateRedemptionState]("DigitalSubscriptionCorporateRedemption", "productType"),
+    Variant.codec[SendThankYouEmailDigitalSubscriptionGiftRedemptionState]("DigitalSubscriptionGiftRedemption", "productType"),
+    Variant.codec[SendThankYouEmailPaperState]("Paper", "productType"),
+    Variant.codec[SendThankYouEmailGuardianWeeklyState]("GuardianWeekly", "productType"),
+  ))
```

Comment:
if we want a particular discriminator, we now have to specify it correctly on every line, to avoid a runtime exeception